File: camase/ukf.py
"""Unscented Kalman filter on the IRW log-price state (model M9)."""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .kalman import H, process_cov, transition


def _chol_psd(P: np.ndarray) -> np.ndarray:
    P = 0.5 * (P + P.T)
    try:
        return np.linalg.cholesky(P)
    except np.linalg.LinAlgError:
        eig, vec = np.linalg.eigh(P)
        eig = np.clip(eig, 1e-18, None)
        return vec @ np.diag(np.sqrt(eig))


@dataclass
class UnscentedIRW:
    """Additive-noise UKF, n=2, 5 sigma points. Same F/Q/H/R as the linear KF."""

    x: np.ndarray = field(default_factory=lambda: np.zeros(2, dtype=np.float64))
    P: np.ndarray = field(default_factory=lambda: np.diag([1e-4, 1e-6]).astype(np.float64))
    alpha: float = 1e-3
    beta: float = 2.0
    kappa: float = 0.0
    last_S: float = 1.0
    last_innov: float = 0.0
    last_nis: float = 1.0

    @property
    def n(self) -> int:
        return 2

    @property
    def n_sigma(self) -> int:
        return 2 * self.n + 1

    def _weights(self) -> tuple[np.ndarray, np.ndarray, float]:
        n = self.n
        lam = self.alpha * self.alpha * (n + self.kappa) - n
        c = n + lam
        wm = np.full(2 * n + 1, 1.0 / (2.0 * c))
        wc = wm.copy()
        wm[0] = lam / c
        wc[0] = lam / c + (1.0 - self.alpha * self.alpha + self.beta)
        return wm, wc, c

    def sigma_points(self) -> np.ndarray:
        wm, wc, c = self._weights()
        col = _chol_psd(c * self.P)
        pts = np.zeros((self.n_sigma, 2))
        pts[0] = self.x
        for i in range(self.n):
            pts[1 + i] = self.x + col[:, i]
            pts[1 + self.n + i] = self.x - col[:, i]
        return pts

    def predict(self, F: np.ndarray, Q: np.ndarray) -> None:
        wm, wc, _ = self._weights()
        sig = self.sigma_points()
        prop = sig @ F.T
        self.x = wm @ prop
        d = prop - self.x
        self.P = (d * wc[:, None]).T @ d + Q
        self.P = 0.5 * (self.P + self.P.T)

    def update(self, y: float, R: float) -> float:
        wm, wc, _ = self._weights()
        sig = self.sigma_points()
        yhat_i = sig @ H.T  # (5, 1)
        yhat = float(wm @ yhat_i.ravel())
        dy = yhat_i.ravel() - yhat
        S = float(np.sum(wc * dy * dy) + R)
        if S <= 1e-18:
            S = 1e-18
        dx = sig - self.x
        Pxy = (dx * wc[:, None]).T @ dy.reshape(-1, 1)
        K = Pxy / S
        innov = y - yhat
        self.x = self.x + K.ravel() * innov
        self.P = self.P - S * (K @ K.T)
        self.P = 0.5 * (self.P + self.P.T)
        self.last_S = S
        self.last_innov = float(innov)
        self.last_nis = float(innov * innov / S)
        return self.last_nis

    def one_step_pred(self, F: np.ndarray) -> float:
        return float((H @ (F @ self.x))[0])


def run_ukf(prices: np.ndarray, R: float, sigma_a2: float, dt: float = 1.0) -> dict:
    """Convenience runner used by tests."""
    y = np.log(np.maximum(prices, 1e-12))
    ukf = UnscentedIRW()
    F = transition(dt)
    Q = process_cov(sigma_a2, dt)
    p = np.empty_like(y)
    for t, yt in enumerate(y):
        ukf.predict(F, Q)
        ukf.update(float(yt), R)
        p[t] = ukf.x[0]
    return {"p_hat": p, "n_sigma": ukf.n_sigma}

File: src/lib/camase/ukf.ts
import { processCov, transition } from "./kalman";

function cholPsd(P: number[][]): number[][] {
  const a = 0.5 * (P[0][0] + P[0][0]);
  const b = 0.5 * (P[0][1] + P[1][0]);
  const c = P[1][1];
  const S = [
    [a, b],
    [b, c],
  ];
  // 2x2 Cholesky
  const l00 = Math.sqrt(Math.max(S[0][0], 1e-18));
  const l10 = S[1][0] / l00;
  const l11 = Math.sqrt(Math.max(S[1][1] - l10 * l10, 1e-18));
  return [
    [l00, 0],
    [l10, l11],
  ];
}

export class UnscentedIRW {
  x: [number, number] = [0, 0];
  P: number[][] = [
    [1e-4, 0],
    [0, 1e-6],
  ];
  alpha = 1e-3;
  beta = 2;
  kappa = 0;
  lastNis = 1;

  get nSigma() {
    return 5;
  }

  weights(): { wm: number[]; wc: number[]; c: number } {
    const n = 2;
    const lam = this.alpha * this.alpha * (n + this.kappa) - n;
    const c = n + lam;
    const wm = Array(5).fill(1 / (2 * c));
    const wc = [...wm];
    wm[0] = lam / c;
    wc[0] = lam / c + (1 - this.alpha * this.alpha + this.beta);
    return { wm, wc, c };
  }

  sigmaPoints(): number[][] {
    const { c } = this.weights();
    const L = cholPsd([
      [c * this.P[0][0], c * this.P[0][1]],
      [c * this.P[1][0], c * this.P[1][1]],
    ]);
    // columns of L
    const c0: [number, number] = [L[0][0], L[1][0]];
    const c1: [number, number] = [L[0][1], L[1][1]];
    return [
      [this.x[0], this.x[1]],
      [this.x[0] + c0[0], this.x[1] + c0[1]],
      [this.x[0] + c1[0], this.x[1] + c1[1]],
      [this.x[0] - c0[0], this.x[1] - c0[1]],
      [this.x[0] - c1[0], this.x[1] - c1[1]],
    ];
  }

  predict(F: number[][], Q: number[][]) {
    const { wm, wc } = this.weights();
    const sig = this.sigmaPoints();
    const prop = sig.map((s) => [F[0][0] * s[0] + F[0][1] * s[1], F[1][0] * s[0] + F[1][1] * s[1]]);
    this.x = [0, 0];
    for (let i = 0; i < 5; i++) {
      this.x[0] += wm[i] * prop[i][0];
      this.x[1] += wm[i] * prop[i][1];
    }
    this.P = [
      [Q[0][0], Q[0][1]],
      [Q[1][0], Q[1][1]],
    ];
    for (let i = 0; i < 5; i++) {
      const d0 = prop[i][0] - this.x[0];
      const d1 = prop[i][1] - this.x[1];
      this.P[0][0] += wc[i] * d0 * d0;
      this.P[0][1] += wc[i] * d0 * d1;
      this.P[1][0] += wc[i] * d1 * d0;
      this.P[1][1] += wc[i] * d1 * d1;
    }
  }

  update(y: number, R: number) {
    const { wm, wc } = this.weights();
    const sig = this.sigmaPoints();
    const yi = sig.map((s) => s[0]);
    let yhat = 0;
    for (let i = 0; i < 5; i++) yhat += wm[i] * yi[i];
    let S = R;
    let Pxy0 = 0;
    let Pxy1 = 0;
    for (let i = 0; i < 5; i++) {
      const dy = yi[i] - yhat;
      S += wc[i] * dy * dy;
      Pxy0 += wc[i] * (sig[i][0] - this.x[0]) * dy;
      Pxy1 += wc[i] * (sig[i][1] - this.x[1]) * dy;
    }
    if (S <= 1e-18) S = 1e-18;
    const K0 = Pxy0 / S;
    const K1 = Pxy1 / S;
    const innov = y - yhat;
    this.x[0] += K0 * innov;
    this.x[1] += K1 * innov;
    this.P[0][0] -= S * K0 * K0;
    this.P[0][1] -= S * K0 * K1;
    this.P[1][0] -= S * K1 * K0;
    this.P[1][1] -= S * K1 * K1;
    this.lastNis = (innov * innov) / S;
    return this.lastNis;
  }

  oneStepPred(F: number[][]) {
    return F[0][0] * this.x[0] + F[0][1] * this.x[1];
  }
}

export { processCov, transition };
